hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotCache.java

```diff
@@ -0,0 +1,375 @@
+      try {
+        // TODO: If wrapped with SoftReference<>, omSnapshot could be null?
+        omSnapshot.close();
+      } catch (IOException e) {
+        throw new IllegalStateException("Failed to close snapshot", e);
+      }
+      it.remove();
+    }
+  }
+
+  /**
+   * State the reason the current thread is getting the OmSnapshot instance.
+   */
+  public enum Reason {
+    FS_API_READ,
+    SNAPDIFF_READ,
+    DEEP_CLEAN_WRITE,
+    GARBAGE_COLLECTION_WRITE
+  }
+
+  public ReferenceCounted<IOmMetadataReader, SnapshotCache> get(String key)
+      throws IOException {
+    return get(key, false);
+  }
+
+  /**
+   * Get or load OmSnapshot. Shall be close()d after use.
+   * TODO: [SNAPSHOT] Can add reason enum to param list later.
+   * @param key snapshot table key
+   * @return an OmSnapshot instance, or null on error
+   */
+  public ReferenceCounted<IOmMetadataReader, SnapshotCache> get(String key,
+      boolean skipActiveCheck) throws IOException {
+    // Atomic operation to initialize the OmSnapshot instance (once) if the key
+    // does not exist.
+    ReferenceCounted<IOmMetadataReader, SnapshotCache> rcOmSnapshot =
+        dbMap.computeIfAbsent(key, k -> {
+          LOG.info("Loading snapshot. Table key: {}", k);
+          try {
+            return new ReferenceCounted<>(cacheLoader.load(k), false, this);
+          } catch (OMException omEx) {
+            // Return null if the snapshot is no longer active
+            if (!omEx.getResult().equals(FILE_NOT_FOUND)) {
+              throw new IllegalStateException(omEx);
+            }
+          } catch (IOException ioEx) {
+            // Failed to load snapshot DB
+            throw new IllegalStateException(ioEx);
+          } catch (Exception ex) {
+            // Unexpected and unknown exception thrown from CacheLoader#load
+            throw new IllegalStateException(ex);
+          }
+          // Do not put the value in the map on exception
+          return null;
+        });
+
+    if (rcOmSnapshot == null) {
+      // The only exception that would fall through the loader logic above
+      // is OMException with FILE_NOT_FOUND.
+      throw new OMException("Snapshot table key '" + key + "' not found, "
+          + "or the snapshot is no longer active",
+          OMException.ResultCodes.FILE_NOT_FOUND);
+    }
+
+    // If the snapshot is already loaded in cache, the check inside the loader
+    // above is ignored. But we would still want to reject all get()s except
+    // when called from SDT (and some) if the snapshot is not active anymore.
+    if (!skipActiveCheck &&
+        !omSnapshotManager.isSnapshotStatus(key, SNAPSHOT_ACTIVE)) {
+      throw new OMException("Unable to load snapshot. " +
+          "Snapshot with table key '" + key + "' is no longer active",
+          FILE_NOT_FOUND);
+    }
+
+    // Increment the reference count on the instance.
+    rcOmSnapshot.incrementRefCount();
```

Comment:
It seems like a problem that the cache load and the ref count inc are not synchronized.

Suppose thread 1 has a ref to snap1 and thread 2 is running get() on snap1.

Then, when thread 2 has passed over the computeIfAbsent() but before the count is incremented, thread 1 swaps in and releases snap1, lowers the count to 0 and cleansup the cache.

Then thread 1 comes back and increments the count.

Now the ref count is 1, but the snapshot isn't in memory at all.

Reply:
Good catch. I've changed `dbMap.computeIfAbsent()` into `dbMap.compute()` and increment ref count inside the atomic lambda block to patch this hole.